<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Проект-по-SQL." data-toc-modified-id="Проект-по-SQL.-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Проект по SQL.</a></span></li><li><span><a href="#Описание-проекта." data-toc-modified-id="Описание-проекта.-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Описание проекта.</a></span></li><li><span><a href="#Описание-данных." data-toc-modified-id="Описание-данных.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Описание данных.</a></span></li><li><span><a href="#Задания." data-toc-modified-id="Задания.-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Задания.</a></span></li><li><span><a href="#Подготовка-к-запросу." data-toc-modified-id="Подготовка-к-запросу.-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Подготовка к запросу.</a></span></li><li><span><a href="#Запросы-к-СУБД." data-toc-modified-id="Запросы-к-СУБД.-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Запросы к СУБД.</a></span><ul class="toc-item"><li><span><a href="#Исследование-таблиц." data-toc-modified-id="Исследование-таблиц.-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Исследование таблиц.</a></span></li><li><span><a href="#Запрос-№1" data-toc-modified-id="Запрос-№1-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Запрос №1</a></span></li><li><span><a href="#Запрос-№2" data-toc-modified-id="Запрос-№2-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Запрос №2</a></span></li><li><span><a href="#Запрос-№3" data-toc-modified-id="Запрос-№3-6.4"><span class="toc-item-num">6.4&nbsp;&nbsp;</span>Запрос №3</a></span></li><li><span><a href="#Запрос-№4" data-toc-modified-id="Запрос-№4-6.5"><span class="toc-item-num">6.5&nbsp;&nbsp;</span>Запрос №4</a></span></li><li><span><a href="#Запрос-№5" data-toc-modified-id="Запрос-№5-6.6"><span class="toc-item-num">6.6&nbsp;&nbsp;</span>Запрос №5</a></span></li></ul></li></ul></div>

## Проект по SQL.

## Описание проекта.

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Описание данных.

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Задания.

*Задания*:

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Подготовка к запросу.

In [1]:
# импортируем библиотеки
!pip install psycopg2
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных

In [3]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, # порт подключения
            'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Запросы к СУБД.

### Исследование таблиц.

Выведем по первые 5 строк каждой из таблиц.

In [4]:
query_books_head = ''' SELECT *
            FROM books
            Limit 5
        '''

In [5]:
pd.io.sql.read_sql(query_books_head, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [6]:
query_authors_head = ''' SELECT *
            FROM authors
            Limit 5
        '''

In [7]:
pd.io.sql.read_sql(query_authors_head, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [8]:
query_ratings_head = ''' SELECT *
            FROM ratings
            Limit 5
        '''

In [9]:
pd.io.sql.read_sql(query_ratings_head, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [10]:
query_reviews_head = ''' SELECT *
            FROM reviews
            Limit 5
        '''

In [11]:
pd.io.sql.read_sql(query_reviews_head, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [12]:
query_publishers_head = ''' SELECT *
            FROM publishers
            Limit 5
        '''

In [13]:
pd.io.sql.read_sql(query_publishers_head, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

### Запрос №1

Посчитайте, сколько книг вышло после 1 января 2000 года;

In [14]:
query_1 = ''' SELECT COUNT(book_id)
            FROM books
            WHERE publication_date > '01.01.2000'
        '''

In [15]:
pd.io.sql.read_sql(query_1, con = engine)

count
0    819

**После 1 января 2000 года вышло 819 книг.**

### Запрос №2

Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [16]:
query_2 = """
SELECT 
    b.title, 
    COUNT(DISTINCT rv.review_id) AS reviews_count, 
    AVG(r.rating) AS average_rating
FROM 
    books b
    LEFT JOIN ratings r ON b.book_id = r.book_id
    LEFT JOIN reviews rv ON b.book_id = rv.book_id
GROUP BY 
    b.book_id
ORDER BY reviews_count DESC;
"""

In [17]:
pd.io.sql.read_sql(query_2, con = engine)

title  reviews_count  \
0                              Twilight (Twilight  #1)              7   
1                                  Water for Elephants              6   
2                                     The Glass Castle              6   
3    Harry Potter and the Prisoner of Azkaban (Harr...              6   
4    The Curious Incident of the Dog in the Night-Time              6   
..                                                 ...            ...   
995    Anne Rice's The Vampire Lestat: A Graphic Novel              0   
996                            The Natural Way to Draw              0   
997   The Cat in the Hat and Other Dr. Seuss Favorites              0   
998                          Essential Tales and Poems              0   
999  Disney's Beauty and the Beast (A Little Golden...              0   

     average_rating  
0          3.662500  
1          3.977273  
2          4.206897  
3          4.414634  
4          4.081081  
..              ...  
995        3.666667  
996        3.000000  
997        5.000000  
998        4.000000  
999        4.000000  

[1000 rows x 3 columns]

**Лидером является книга Twilight#1 на неё написано 7 обзоров, средняя оценка 3.66**

### Запрос №3

Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [18]:
query_3 = ''' 
SELECT 
    top.publisher_id,
    top.publisher,
    top.number_books
FROM 
    (SELECT
         b.publisher_id AS publisher_id,
         COUNT(b.book_id) AS number_books,
         p.publisher AS publisher
     FROM 
         books AS b
         LEFT JOIN publishers AS p ON b.publisher_id = p.publisher_id
     WHERE 
         num_pages > 50
     GROUP BY 
         b.publisher_id,
         p.publisher
     ORDER BY    
             number_books DESC) AS top
LIMIT 1
; '''

In [19]:
pd.io.sql.read_sql(query_3, con = engine)

publisher_id      publisher  number_books
0           212  Penguin Books            42

**Издательство Penguin Books выпустило 42 книги толщиной более 50 страниц.**

### Запрос №4

Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [20]:
query_4 = '''
SELECT author_ratings.author,
       AVG(author_ratings.avg_rating) AS total_rating
FROM
    (SELECT a.author AS author,
            AVG(ra.rating) AS avg_rating
     FROM books AS b
     LEFT JOIN authors AS a ON b.author_id = a.author_id
     LEFT JOIN ratings AS ra ON b.book_id = ra.book_id
     GROUP BY b.book_id,
              a.author_id
     HAVING COUNT(ra.rating_id) > 50
     ) AS author_ratings
GROUP BY author_ratings.author
ORDER BY total_rating DESC
LIMIT 1
; '''

In [21]:
pd.io.sql.read_sql(query_4, con = engine)

author  total_rating
0  J.K. Rowling/Mary GrandPré      4.283844

**J.K. Rowling/Mary GrandPré - автор с самой высокой средней оценкой 4.283844.**

### Запрос №5

Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [22]:
query_5 = '''
SELECT AVG(users.total_reviews)

FROM
(SELECT username,
        COUNT(review_id) AS total_reviews 
 FROM reviews
 WHERE username IN (
                    SELECT username
                    FROM ratings
                    GROUP BY username
                    HAVING COUNT(rating_id) > 50
                    )
 GROUP BY username
 ) AS users

; '''

In [23]:
pd.io.sql.read_sql(query_5, con = engine)

avg
0  24.333333

**24.33 - среднее количество обзоров от пользователей, которые поставили больше 50 оценок.**